<a href="https://colab.research.google.com/github/ZhengXiaohu98/448Assignment/blob/main/Assignment4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Xiaohu Zheng (Assigment 4)

In [14]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split # need to splitting datasets
# some built-in models
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
# import the dataset
from sklearn.datasets import load_iris

# Part 1: Load the dataset

In [15]:
# Load the dataset (load remotely, not locally)
dataset = load_iris()
dataframe = pd.DataFrame(data=dataset.data, columns=dataset.feature_names)
dataframe["lable"] = dataset.target
# Output the first 15 rows of the data
dataframe.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    lable  
0       0  
1       0  
2       0  
3       0  
4       0  
5       0  
6       0  
7       0  
8       0  
9       0  
10      0  
11      0  
12      0  
13      0  
14      0

In [16]:
# Display a summary of the table information (number of datapoints, etc.)
dataframe.info()
dataframe.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   lable              150 non-null    int32  
dtypes: float64(4), int32(1)
memory usage: 5.4 KB


sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)       lable  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?
- The first we can see that we have 150 samples in our iris dartaset.
- For features we have sepal length, sepal width, petal length, petal width, so in total we have 4 features.
- We only have one lable, and its value is between 0 - 2, 0 represents `Iris-Setosa`, 1 represents `Iris-Versicolour`, 2 represents `Iris-Virginica`.

# Part 2: Split the dataset into train and test

In [34]:
# Take the dataset and split it into our features (X) and label (y)
X = dataframe[['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']]
y = dataframe['lable']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

# Part 3: Logistic Regression

In [36]:
# i. Use sklearn to train a LogisticRegression model on the training set
lg_model = LogisticRegression(max_iter=500).fit(X_train, y_train)

In [37]:
# ii. For a sample datapoint, predict the probabilities for each possible class
x_sample_point = np.array([[5.,4.,1.5,0.2]])
y_sample_pred = lg_model.predict_proba(x_sample_point)[0]

print(f"The probability of this sample being Iris-Setosa is: {y_sample_pred[0]:.5f}")
print(f"The probability of this sample being Iris-Versicolour is: {y_sample_pred[1]:.5f}")
print(f"The probability of this sample being Iris-Virginica is: {y_sample_pred[2]:.5f}")

The probability of this sample being Iris-Setosa is: 0.98654
The probability of this sample being Iris-Versicolour is: 0.01346
The probability of this sample being Iris-Virginica is: 0.00000


C:\Users\xiaoh\miniconda3\envs\appliedml\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


In [39]:
# iii. Report on the score for Logistic regression model, what does the score measure?
train_score = lg_model.score(X_train, y_train)
test_score = lg_model.score(X_test, y_test)
print(f"The (train, test) score for our logistic regression model is ({train_score:.5f}, {test_score:.5f}).")

The (train, test) score for our logistic regression model is (0.97778, 1.00000).


- The score measures how good our model's performance
- The training score mesures our model's performance on the tranning set. We have a value of 0.97778, which is very good.
- The testing socre mesures our model's performance on the real world data, it is 1.0, which means we have a perfect model.
- The testing socre is higher than the training score which is unexpected.

In [44]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
w = np.round(lg_model.coef_, decimals=5)
b = np.round(lg_model.intercept_, 5)
print("The coefficents of our model is: ", w)
print("The intercept of our model is: ", b)

The coefficents of our model is:  [[-0.42669  0.97251 -2.44451 -1.03217]
 [ 0.51317 -0.22361 -0.21517 -0.85144]
 [-0.08648 -0.7489   2.65968  1.8836 ]]
The intercept of our model is:  [  9.50115   1.91149 -11.41264]


# Part 4: Support Vector Machine

In [46]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svm_model = SVC(probability=True, random_state=42).fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
x_sample_point = np.array([[5.,4.,1.5,0.2]])
y_sample_pred = svm_model.predict_proba(x_sample_point)[0]

print(f"The probability of this sample being Iris-Setosa is: {y_sample_pred[0]:.5f}")
print(f"The probability of this sample being Iris-Versicolour is: {y_sample_pred[1]:.5f}")
print(f"The probability of this sample being Iris-Virginica is: {y_sample_pred[2]:.5f}")

The probability of this sample being Iris-Setosa is: 0.97742
The probability of this sample being Iris-Versicolour is: 0.01284
The probability of this sample being Iris-Virginica is: 0.00974


C:\Users\xiaoh\miniconda3\envs\appliedml\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(


In [47]:
# iii. Report on the score for the SVM, what does the score measure?
train_score = svm_model.score(X_train, y_train)
test_score = svm_model.score(X_test, y_test)
print(f"The (train, test) score for our logistic regression model is ({train_score:.5f}, {test_score:.5f}).")

The (train, test) score for our logistic regression model is (0.97037, 1.00000).


- The score mesures how good our SVM model is, we have a socre 0.97037 for training set, and 1 for testing set. Both the results are pretty good. Especially the testing socre has value of one, which means we have a perfect model.

# Part 5: Neural Network

In [52]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp_model = MLPClassifier(max_iter=100, random_state=42).fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
x_sample_point = np.array([[5.,4.,1.5,0.2]])
y_sample_pred = svm_model.predict_proba(x_sample_point)[0]

print(f"The probability of this sample being Iris-Setosa is: {y_sample_pred[0]:.5f}")
print(f"The probability of this sample being Iris-Versicolour is: {y_sample_pred[1]:.5f}")
print(f"The probability of this sample being Iris-Virginica is: {y_sample_pred[2]:.5f}")


The probability of this sample being Iris-Setosa is: 0.97742
The probability of this sample being Iris-Versicolour is: 0.01284
The probability of this sample being Iris-Virginica is: 0.00974


C:\Users\xiaoh\miniconda3\envs\appliedml\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\xiaoh\miniconda3\envs\appliedml\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(


In [53]:
# iii. Report on the score for the Neural Network, what does the score measure?

train_score = mlp_model.score(X_train, y_train)
test_score = mlp_model.score(X_test, y_test)
print(f"The (train, test) score for our logistic regression model is ({train_score:.5f}, {test_score:.5f}).")

The (train, test) score for our logistic regression model is (0.92593, 0.86667).


- The socre measures the model's performance. We have a training score of 0.92593, it a good number for just 100 iterations, and we have a testing score of 0.86667, which means the model will have 86% accuracy on real world data.

In [54]:
# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)
mlp_model = MLPClassifier(max_iter=1000, random_state=42).fit(X_train, y_train)

train_score = mlp_model.score(X_train, y_train)
test_score = mlp_model.score(X_test, y_test)
print(f"The (train, test) score for our logistic regression model is ({train_score:.5f}, {test_score:.5f}).")

The (train, test) score for our logistic regression model is (0.98519, 1.00000).


- We can simply just increase the epoches because when the epoches is 100, it seems like it didn't fully converge yet
- This time we get a train score of 0.98519, and a test score of 1. So our model is pretty good now, and it can be considered as a perfect model since we have a testing accuracy of 1.

# Part 6: K-Nearest Neighbors

In [55]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
knn_model = KNeighborsClassifier().fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
# ii. For a sample datapoint, predict the probabilities for each possible class
x_sample_point = np.array([[5.,4.,1.5,0.2]])
y_sample_pred = knn_model.predict_proba(x_sample_point)[0]

print(f"The probability of this sample being Iris-Setosa is: {y_sample_pred[0]:.5f}")
print(f"The probability of this sample being Iris-Versicolour is: {y_sample_pred[1]:.5f}")
print(f"The probability of this sample being Iris-Virginica is: {y_sample_pred[2]:.5f}")

The probability of this sample being Iris-Setosa is: 1.00000
The probability of this sample being Iris-Versicolour is: 0.00000
The probability of this sample being Iris-Virginica is: 0.00000


C:\Users\xiaoh\miniconda3\envs\appliedml\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(


In [57]:
# iii. Report on the score for kNN, what does the score measure?
train_score = knn_model.score(X_train, y_train)
test_score = knn_model.score(X_test, y_test)
print(f"The (train, test) score for our logistic regression model is ({train_score:.5f}, {test_score:.5f}).")

The (train, test) score for our logistic regression model is (0.97037, 1.00000).


- The train score is 0.97037, which is slighly less than the neural network, but it is still a good result.
- The testing socre is 1, which means our model works perfect for real world data

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

- This notebook tested several different models on the iris dataset. All of them have good result, but the best model is our `Neural Network` model, which is the MLPClassifier. bBecause all the model have the testing score of 1, so it is hard to tell which one is better. Howevery, for the training score, MLP classifier has the best result, which is why I think it performs better than other models. What surprised me is that all the models have testing socre better than the training score, which is rare. The reason might is the dataset is pretty small which has only 150 samples.